lexer/stateful/stateful.go

```diff
@@ -167,45 +335,58 @@ func New(rules Rules) (*Definition, error) {
 					return nil, fmt.Errorf("%s.%d: %s", key, i, err)
 				}
 			}
-			compiled[key] = append(compiled[key], compiledRule{
+
+			compiled[key].rules = append(compiled[key].rules, compiledRule{
 				Rule:   rule,
 				ignore: len(rule.Name) > 0 && unicode.IsLower(rune(rule.Name[0])),
 				RE:     re,
 			})
 		}
 	}
 restart:
-	for state, rules := range compiled {
-		for i, rule := range rules {
+	for state, group := range compiled {
+		for i, rule := range group.rules {
 			if action, ok := rule.Action.(RulesAction); ok {
 				if err := action.applyRules(state, i, compiled); err != nil {
 					return nil, fmt.Errorf("%s.%d: %s", state, i, err)
 				}
-				goto restart
+				continue restart
```

Comment:
This goto is intentional (it's why the label is named `restart`), please revert.